File: WhereAmINowServer/src/main/webapp/js/holders/PropertiesHolder.js
/**
 * Created 2/10/17.
 */
function PropertiesHolder(main) {

    this.type = "properties";
    this.saveable = true;

    var dialog;

    this.start = function() {
        dialog = u.dialog({
            title: u.lang.set_my_name,
            items: [
                { type: HTML.INPUT, label: u.lang.name }
            ],
            positive: {
                label: u.lang.ok,
                onclick: function(args) {
                    if(args[0].value) {
                        var name = args[0].value;
                        u.save("properties:name", name);
                        u.save("properties:name_asked", true);
                        main.me.fire(EVENTS.CHANGE_NAME, name);
                    }
                    console.log(args);
                }
            },
            negative: {
                label: "Cancel"
            }
        });
    };

    this.onEvent = function(EVENT,object){
        var self = this;
        switch (EVENT){
            case EVENTS.TRACKING_ACTIVE:
                if(!u.load("properties:name") && !u.load("properties:name_asked")) {
                    var askIfNameNotDefinedDialog = u.dialog({
                        items: [
                            { type: HTML.DIV, label: u.lang.your_name_is_not_defined }
                        ],
                        positive: {
                            label: u.lang.yes,
                            onclick: function(args) {
                                setMyName.call(main.me);
                            }
                        },
                        negative: {
                            label: u.lang.no,
                            onclick: function(){
                                u.save("properties:name_asked", true);
                            }
                        },
                        neutral: {
                            label: u.lang.maybe_later
                        },
                        timeout: 10000
                    }).open();
                }
                var name = main.me.name;
                if(!name && main.me.properties) name = main.me.properties.name;
                if(name) {
                    main.me.fire(EVENTS.CHANGE_NAME, name);
                }
                break;
            case EVENTS.CREATE_CONTEXT_MENU:
                var user = this;
                if(user.number == main.me.number) {
                    object.add(MENU.SECTION_PRIMARY, self.type + "_set_my_name", u.lang.set_my_name, "face", function () {
                        setMyName.call(user);
                    });
                }
                break;
            case EVENTS.SELECT_USER:
                this.properties.selected = true;
                break;
            case EVENTS.UNSELECT_USER:
                this.properties.selected = false;
                break;
            case EVENTS.SELECT_SINGLE_USER:
                main.users.forAllUsers(function(number,user){
                    user.properties.selected = false;
                });
                var myUser = this;
                myUser.fire(EVENTS.SELECT_USER);
                main.users.forAllUsers(function(number,user){
                    if(myUser != user) {
                        user.fire(EVENTS.UNSELECT_USER);
                    }
                });
                break;
            case EVENTS.CHANGE_NAME:
                if(this.properties)this.properties.name = object;
                if(this.number == main.me.number) {
                    main.tracking.put(USER.NAME, object);
                    main.tracking.send(REQUEST.CHANGE_NAME);
                }
                break;
            case EVENTS.CHANGE_NUMBER:
                if(this.properties)this.properties.number = object;
                break;
            case EVENTS.CHANGE_COLOR:
                if(this.properties)this.properties.color = object;
                break;
            case EVENTS.MAKE_ACTIVE:
                if(this.properties)this.properties.active = true;
                break;
            case EVENTS.MAKE_INACTIVE:
                if(this.properties)this.properties.active = false;
                break;
            case EVENTS.MAP_READY:
                main.me.createViews();
                main.me.fire(EVENTS.SELECT_USER);
                main.me.fire(EVENTS.CHANGE_COLOR, "#0000FF");
                main.me.fire(EVENTS.MAKE_ACTIVE);
                break;
            default:
                break;
        }
        return true;
    };

    this.createView = function(myUser) {
        var view = {
            user: myUser,
            color: myUser.color,
            name: myUser.name,
            number: myUser.number,
            active: myUser.active,
            selected: myUser.selected,
            getDisplayName: getDisplayName.bind(myUser),
        };

        delete myUser.color;
        delete myUser.name;
        delete myUser.active;
        delete myUser.selected;
        myUser.properties = view;
        return view;
    };

    function getDisplayName(){
        var name = this.properties.name;
        if(!name){
            if(this.number == main.me.number) {
                name = "Me";
            } else {
                name = "Friend "+this.number;
            }
        }
        return name;
    }

    function setMyName(name){
        if(dialog) dialog.close();
        dialog.items[0].value = main.me.properties.name || "";
        dialog.open();
    }

}

File: app/src/main/java/com/edeqa/waytous/holders/property/PropertiesHolder.java
package com.edeqa.waytous.holders.property;

import android.content.Context;
import android.content.SharedPreferences;
import android.location.Location;

import com.edeqa.waytous.R;
import com.edeqa.waytous.State;
import com.edeqa.waytous.abstracts.AbstractProperty;
import com.edeqa.waytous.abstracts.AbstractPropertyHolder;
import com.edeqa.waytous.helpers.MyUser;
import com.edeqa.waytous.helpers.Utils;
import com.edeqa.waytous.interfaces.Runnable2;
import com.google.maps.android.SphericalUtil;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import static android.content.Context.MODE_PRIVATE;
import static com.edeqa.waytous.Constants.REQUEST_CHANGE_NAME;
import static com.edeqa.waytous.Constants.USER_DISMISSED;
import static com.edeqa.waytous.Constants.USER_JOINED;
import static com.edeqa.waytous.Constants.USER_NAME;
import static com.edeqa.waytous.helpers.Events.ACTIVITY_PAUSE;
import static com.edeqa.waytous.helpers.Events.ACTIVITY_RESUME;
import static com.edeqa.waytous.helpers.Events.CHANGE_COLOR;
import static com.edeqa.waytous.helpers.Events.CHANGE_NAME;
import static com.edeqa.waytous.helpers.Events.CHANGE_NUMBER;
import static com.edeqa.waytous.helpers.Events.MAKE_ACTIVE;
import static com.edeqa.waytous.helpers.Events.MAKE_DISABLED;
import static com.edeqa.waytous.helpers.Events.MAKE_ENABLED;
import static com.edeqa.waytous.helpers.Events.MAKE_INACTIVE;
import static com.edeqa.waytous.helpers.Events.MOVING_AWAY_FROM;
import static com.edeqa.waytous.helpers.Events.MOVING_CLOSE_TO;
import static com.edeqa.waytous.helpers.Events.SELECT_SINGLE_USER;
import static com.edeqa.waytous.helpers.Events.SELECT_USER;
import static com.edeqa.waytous.helpers.Events.TRACKING_ACTIVE;
import static com.edeqa.waytous.helpers.Events.TRACKING_STOP;
import static com.edeqa.waytous.helpers.Events.UNSELECT_USER;

/**
 * Created 11/18/16.
 */
@SuppressWarnings("unchecked")
public class PropertiesHolder extends AbstractPropertyHolder {

    public static final String TYPE = "properties"; //NON-NLS

    public static final String PREFERENCE_MY_NAME = "my_name"; //NON-NLS

    private static final String SELECTED = "selected"; //NON-NLS
    private static final String IMAGE_RESOURCE = "image_resource"; //NON-NLS
    private static final int DISTANCE_MOVING_CLOSE = 50;
    private static final int DISTANCE_MOVING_AWAY = 100;
    private final SharedPreferences sharedPreferences;
    private final Context context;
    private HashMap<String, Serializable> external = new HashMap<>();

    public PropertiesHolder(Context context){
        super(context);
        this.context = context;
        sharedPreferences = context.getSharedPreferences("tracking_active", MODE_PRIVATE); //NON-NLS
    }

    @Override
    public String getType(){
        return TYPE;
    }

    @Override
    public boolean dependsOnEvent() {
        return true;
    }

    @Override
    public boolean onEvent(String event, Object object) {
        switch(event){
            case TRACKING_ACTIVE:
                State.getInstance().getUsers().forAllUsers(new Runnable2<Integer, MyUser>() {
                    @Override
                    public void call(Integer number, MyUser myUser) {
                        HashMap<String, Serializable> value = (HashMap<String, Serializable>) myUser.getProperties().loadFor(TYPE);
                        if(value != null) {
                            if(value.containsKey(SELECTED)) {
                                if(value.get(SELECTED) != null && (boolean)value.get(SELECTED)) {
                                    myUser.getProperties().selected = true;
//                                    myUser.fire(SELECT_USER, 0);
                                }
                            }
                            if(value.containsKey(IMAGE_RESOURCE)) {
                                if(value.get(IMAGE_RESOURCE) != null) {
                                    myUser.getProperties().imageResource = (int) value.get(IMAGE_RESOURCE);
//                                    myUser.getProperties().setButtonView((int) value.get(IMAGE_RESOURCE));
                                }
                            }
                        }
                    }
                });
                /*State.getInstance().getUsers().forAllUsers(new MyUsers.Callback() {
                    @Override
                    public void call(Integer number, MyUser myUser) {
                        HashMap<String, Serializable> value = (HashMap<String, Serializable>) myUser.getProperties().loadFor(TYPE);
                        if(value != null) {
                            if(value.containsKey(SELECTED)) {
                                if(value.get(SELECTED) != null && !(boolean)value.get(SELECTED)) {
                                    myUser.fire(UNSELECT_USER, 0);
                                }
                            }
                        }
                    }
                });*/

                break;
            case ACTIVITY_PAUSE:
                State.getInstance().getUsers().forAllUsers(new Runnable2<Integer, MyUser>() {
                    @Override
                    public void call(Integer number, MyUser myUser) {
                        HashMap<String, Serializable> m = new HashMap<>();
                        m.put(SELECTED, myUser.getProperties().isSelected());
                        m.put(IMAGE_RESOURCE, myUser.getProperties().getImageResource());
                        myUser.getProperties().saveFor(TYPE,m);
                    }
                });
                break;
            case ACTIVITY_RESUME:
                State.getInstance().getUsers().forAllUsers(new Runnable2<Integer, MyUser>() {
                    @Override
                    public void call(Integer number, MyUser myUser) {
                        if(myUser.getProperties().selected) {
                            myUser.fire(SELECT_USER);
                        }
                    }
                });

                break;
            case TRACKING_STOP:
                sharedPreferences.edit().clear().apply();
                break;
            case USER_JOINED:
                MyUser myUser = (MyUser) object;
                myUser.createViews();
                break;
            case USER_DISMISSED:
                myUser = (MyUser) object;
                myUser.removeViews();
                break;

        }
        return true;
    }

    public void saveFor(String type, Serializable props) {
        if (props == null) {
            external.remove(type);
            sharedPreferences.edit().remove(type).apply();
        } else {
            external.put(type, props);
            sharedPreferences.edit().putString(type, Utils.serializeToString(props)).apply();
        }
    }

    public Serializable loadFor(String type){
        if(external.containsKey(type)){
            return external.get(type);
        } else {
            String saved = sharedPreferences.getString(type, null);
            if(saved != null){
                return (Serializable) Utils.deserializeFromString(saved);
            }
        }
        return null;
    }

    @Override
    public Properties create(MyUser myUser) {
        if (myUser == null) return null;
        return new Properties(myUser);
    }

    public class Properties extends AbstractProperty {
        private HashMap<String, Serializable> external = new HashMap<>();

        private String name;
        private String description;
        private int number;
        private int color;
        private int imageResource;
        private long changed;
        private boolean selected;
        private boolean active;
        private double previousDistance;

        Properties(MyUser myUser) {
            super(PropertiesHolder.this.context, myUser);
        }

        @Override
        public String getType() {
            return PropertiesHolder.this.getType();
        }

        @Override
        public boolean dependsOnLocation(){
            return true;
        }

        @Override
        public List<String> events() {
            List<String> list = new ArrayList<>();
            list.add(CHANGE_COLOR);
            return list;
        }

        @Override
        public void onChangeLocation(final Location location) {
            if(location != null && myUser.getProperties().isActive()) {
                new Thread(new Runnable() {
                    @Override
                    public void run() {
                        if(myUser == State.getInstance().getMe()) {
                            State.getInstance().getUsers().forAllUsersExceptMe(new Runnable2<Integer, MyUser>() {
                                @Override
                                public void call(Integer number, MyUser user) {
                                    if(user.getProperties().isActive()) {
                                        checkDistance(myUser, user);
                                    }
                                }
                            });
                        } else {
                            checkDistance(State.getInstance().getMe(), myUser);
                        }
                    }
                }).start();
            }
        }

        @Override
        public boolean onEvent(String event, Object object) {
            switch(event){
                case SELECT_USER:
                    selected = true;
                    break;
                case SELECT_SINGLE_USER:
                    myUser.getProperties().selected = true;
                    final ArrayList<MyUser> unselected = new ArrayList<>();
                    State.getInstance().getUsers().forAllUsers(new Runnable2<Integer, MyUser>() {
                        @Override
                        public void call(Integer number, MyUser user) {
                            if(user != myUser && user.getProperties().isSelected()) {
                                user.getProperties().selected = false;
                                unselected.add(user);
                            }
                        }
                    });
                    myUser.fire(SELECT_USER);
                    for(MyUser user:unselected) {
                        user.fire(UNSELECT_USER);
                    }
                    break;
                case UNSELECT_USER:
                    selected = false;
                    break;
                case MAKE_ACTIVE:
                    active = true;
                    break;
                case MAKE_INACTIVE:
                    active = false;
                    selected = false;
                    break;
                case MAKE_ENABLED:
                case MAKE_DISABLED:
                    changed = (long) object;
                    break;
                case CHANGE_NUMBER:
                    number = (int) object;
                    break;
                case CHANGE_COLOR:
                    color = (int) object;
                    break;
                case CHANGE_NAME:
                    name = (String) object;
                    if(myUser == State.getInstance().getMe()){
                        State.getInstance().setPreference(PREFERENCE_MY_NAME, name);
                        if(State.getInstance().getTracking() != null){
                            if(name == null) name = "";
                            State.getInstance().getTracking().put(USER_NAME,name).send(REQUEST_CHANGE_NAME);
                        }
                    }
                    break;
//                case CHANGE_TYPE:
//                    type = (String) object;
//                    break;
            }
            return true;
        }

        public void setName(String name) {
            this.name = name;
        }

        public String getName() {
            return name;
        }

        public String getDisplayName(){
            String res = name;
            if(name == null || name.length()==0 || "null".equals(name)){
                if(myUser == State.getInstance().getMe()){
                    res = context.getString(R.string.me);
                } else if (getNumber() == 0) {
                    res = context.getString(R.string.leader);
                } else {
                    res = context.getString(R.string.friend_d, getNumber());
                }
            }
            return res;
        }

        public int getNumber() {
            return number;
        }

        public void setNumber(int number) {
            this.number = number;
        }

        public boolean isSelected() {
            return selected;
        }

        public boolean isActive() {
            return active;
        }

        public void setColor(int color) {
            this.color = color;
        }

        public int getColor() {
            return color;
        }

        public void saveFor(String type, Serializable props) {
            if(props == null) {
                external.remove(type);
                sharedPreferences.edit().remove(type + "_" + myUser.getProperties().getNumber()).apply();
            } else {
                external.put(type, props);
                sharedPreferences.edit().putString(type + "_" + myUser.getProperties().getNumber(), Utils.serializeToString(props)).apply();
            }
        }

        public Serializable loadFor(String type){
            if(external.containsKey(type)){
                return external.get(type);
            } else {
                String saved = sharedPreferences.getString(type + "_" + myUser.getProperties().getNumber(), null);
                if(saved != null){
                    return (Serializable) Utils.deserializeFromString(saved);
                }
            }
            return null;
        }

        public void setButtonView(int imageResource) {
            this.imageResource = imageResource;
        }

        public int getImageResource() {
            return imageResource;
        }

        private void checkDistance(MyUser me, MyUser user) {
            if(me.getLocation() == null || user.getLocation() == null) return;
            try {
                double distance = SphericalUtil.computeDistanceBetween(Utils.latLng(me.getLocation()), Utils.latLng(user.getLocation()));
                if (distance <= DISTANCE_MOVING_CLOSE && user.getProperties().previousDistance > DISTANCE_MOVING_CLOSE) {
                    user.fire(MOVING_CLOSE_TO);
                } else if (distance > DISTANCE_MOVING_AWAY && user.getProperties().previousDistance > 0 && user.getProperties().previousDistance < DISTANCE_MOVING_AWAY) {
                    user.fire(MOVING_AWAY_FROM);
                }
                user.getProperties().previousDistance = distance;
            } catch (Exception e){
                e.printStackTrace();
            }
        }

        @SuppressWarnings("HardCodedStringLiteral")
        @Override
        public String toString() {
            return "Properties{" +
                    "name='" + name + '\'' +
                    ", number=" + number +
                    ", color=" + color +
                    ", selected=" + selected +
                    ", active=" + active +
                    '}';
        }

        public String getDescription() {
            return description;
        }

        public void setDescription(String description) {
            this.description = description;
        }

        public long getChanged() {
            return changed;
        }

        public void setChanged(long changed) {
            this.changed = changed;
        }
    }

}
